### ISSUE TITLE
Soap WS integration example

### ISSUE THREAD

**Issue by marioestradarosa** [strongloop/loopback-next#1550]
## Feature proposal
A Soap Web Service integration example and tutorial.

## Current Behavior
There is only one example for REST service integration.

## Expected Behavior
To have full documentation on all possible scenarios for new app developers.

## What I did so far?
I created a sample application here in my repository [ Soap WS sample ](https://github.com/marioestradarosa/loopback-soap-example)  after a lot of research, so it would be important to make things easier for developers.

## Question 
Do you think is feasible that I write the tutorial and this example be integrated in the examples of the main repo ? of course after writing the test and follow the tutorial formats ?  

What do you think @bajtos  , @virkt25 , @shimks , @dhmlau  et all ?

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

**Comment by virkt25** [comment]
Hi @marioestradarosa, 

I checked out the `Soap WS sample` and it'd be great to have it listed as an example with some tweaks. For example in the tutorial in certain places, `@loopback/cli` can be used to generate some of the artifacts. Let me know if you need any help with it. I'd also let @bajtos, @dhmlau chime in before starting work on it though but a 👍 from me.

I noticed the tutorial mentions a bug in the Dependency Injection framework causing a 500 for the first call to the service? Is this still an issue you're facing -- I didn't find an open issue for that in the `loopback-next` repository.

**Comment by marioestradarosa** [comment]
Hi @virkt25 ,

Yes, that 500 error is coming along every time you access one of the end points for the first time. It complaints that that method for that service don't exist. And then in the second hit it seems like it is available, so it seems the class is not being instantiated? or something. 

Yes, that is  one of the tutorials I am using internally as we are moving migrating and learning the framework. For the SOAP methods, even if you can place them as a string in the wizard, I didn't mentioned it because my WS require more work, but of course can be placed there. Something I noticed is that @loopback/cli  should have installed the service-proxy automatically, but it didn't at least in my exercise, that's why it is an extra step.

And yes, please help me with the CLI options I might be missing! . When agreed,  I will start working on the example to fit @loopback-next requirements and include the tutorial in the documentation.

**Comment by bajtos** [comment]
@marioestradarosa thank you for sharing your tutorial with us, it would be great to have it as part of our official docs.

I'd like to ask you to make few changes and improvements:
 - The tutorial text needs to be in Markdown, not PDF (sorry if this was already obvious).
 - There should be one acceptance test for each REST API method.
 - It would be nice to have integration tests to verify our calculator service client works as expected.
 - Please check the OpenAPI spec generated for your tutorial server - I suspect request bodies won't be correctly described. I think you need to add `@model` and `@property` decorators to your input type. I think @jannyHou knows most about this part.
 - I am proposing to change the REST API so that input numbers are provided via URL query string, I consider such API as more REST-like and easier to use. It will also allows us to change the request method to `GET`, which is the first step to enable caching of responses. (`1+2` is always going to return `3`, no need to recalculate it on every request.)

Thoughts?

**Comment by bajtos** [comment]
> I created a sample application here in my repository Soap WS sample after a lot of research, so it would be important to make things easier for developers.

Can you still remember which parts were not clear and required most research? It would be great if we could improve our documentation and provide the necessary information right in the place where new LoopBack 4 developers are going to look for it.

**Comment by marioestradarosa** [comment]
HI @bajtos ,

The "A lot of research" _did not involve_ **LoopBack Next** the whole time, because we have some web services that are not documented and only can get information from them either from their wsdl doc or using the describe method from the soap module itself, so sorry to not explain that.  

But, let me explain what would be the flow a programmer assigned (with zero/little loopback experience)  this SOAP integration task, based upon exercises we performed here internally

1. He goes into the examples first, and the Todo list has a hidden perl there (the REST integration), so it was not outstanding initially but that served us as a good starting point. He would look for a concise and simple example about the task he's been assigned.

2. He correlates the section "**Calling other APIS**" with the code in the Todo Geo Point query, to asking himself how to map those methods to SOAP ?. So, he would have to go to the documentation for loopback-connector-soap (_He had to know about its existence first right?_) , and see the differences in the operations section for REST and for SOAP by comparing samples there and match with the GeoPoint Rest Service. It was so obvious from that point how they (REST and SOAP) were different in terms of configuration.

3. Our SOAP WS, require authentication, by sending an authentication object in the SoapHeaders, so it was nice documented also in loopback-connector-soap

So, remember there is a potential market for **loopback-next** in the **integration** arena, so we will have application developers that would like a fast an easy way to integrate their soap, rest, dbs, emails, fs etc. to a nice Rest Apis without knowing the details sometimes :-) . The success for Loopback-Next adoption in this group of developers is the examples and documentation for specific cases, unless of course the company had made the decision already and they are forced to do it in loopback-next !

**Input Parameters?**
I agree with you, but could it be possible make two types of endpoints ?, one for Query parameters and the other for the Body parameter?.  for the body parameter as it is at this moment was to ensure the following body can be received by the controller to make a call to a service method. 

```javascript
ReadAccountRq : {
           accountKeys : {
                          standardAccountId : "",
                          pseudonym : {
                                  pseudonymId : '',
                                  pseudonymType : '',
                                  isoCurrencyCode : '',
                                  branchCode : '',
                                  contextType : '',
                                  contextValue : ''
                          },
                          inputAccount : {
                                inputAccountId : '',
                                accountFormatType : ''
                          }
          } 
```

Thank you @bajtos  for assigning me this task, I will do my best to have it done rapidly.

**Comment by shimks** [comment]
Closing as done (#1560)
